core/Pos/Core/Common/Types.hs

```diff
@@ -287,15 +288,16 @@ instance Bounded Coin where
 maxCoinVal :: Word64
 maxCoinVal = 45000000000000000
 
--- | Make Coin from Word64.
+-- | Just use the 'Coin' constructor. This is legacy but still used. Formerly
```

Comment:
I vote against this change. `mkCoin` was designed to use inside the library, so any place where it could fail is a legitimate `error`. Why do not have both smart constructors and verification? Sometimes `mkX` is as easy as `mkX v = build v >> checkX` (sometimes it's not), but it doesn't mean `mkX` doesn't make sense. `checkCoin` is cheap, not noticing a bug can cost us million of dollars.

Reply:
I'm not super opinionated on this either way.  It may be a good idea to put the check and `error` back in. But it would be interesting to discuss the possible catastrophes that could happen due to a `Coin` that exceeds `maxCoinVal`. Could it lead to a million dollar bug?

Reply:
Overflow and subtracting random sum instead of adding non-random?

Reply:
I think we can forget about this one. We'll deal with this problem in CSL-2072 (verification, done by me) and CSL-2173 (unsafe `error` calls etc, done by Vlad int-index).

Reply:
I'd leave this check since it's cheap and adds some safety (we don't know which bug our code can have until they are discovered 🤷‍♂️), but I won't be super disappointed if this check is removed.

Anyway, what I see now is a mess. I see such diff:
```
 mkCoin :: Word64 -> Coin
 mkCoin c
     | c <= maxCoinVal = Coin c
-    | otherwise       = error $ "mkCoin: " <> show c <> " is too large"
-{-# INLINE mkCoin #-}
+    | otherwise = error $ "Coin: " <> show c <> " is too large"
+
+checkCoin :: MonadError Text m => Coin -> m ()
+checkCoin (Coin c)
+    | c <= maxCoinVal = pure ()
+    | otherwise       = throwError $ "Coin: " <> show c <> " is too large"
```

Apparently you changed your decision, but didn't revert changes properly 🤷‍♂️ . Specifically I don't like the following:
1. The check is duplicated (it's in `mkCoin` and in `checkCoin`), I think it's better to reuse `checkCoin` in `mkCoin`.
2. You removed `INLINE`, was it intentional?
3. The comment is outdated.